### ISSUE TITLE
Complement the .join() function with an outer_join(), inner_join(), full/symmetric_join()

### ISSUE THREAD

**Issue by bananenpampe** [lab-cosmo/equistore#315]
Currently, joining requires tensor maps with identical keys (I hope this is still the case?).

It would be nice to have a join function that joins along and returns tensor maps only with the intersecting keys,
or adds the left and right-handed differences of the two keys and blocks to the new TensorMap.

This is relevant to writing a torch data loader for structures with different compositions.
A general collate_fn in a Dataloader can then use the join function, even if the TensorMaps from the torch Dataset have different keys due to other species pairs or compositions.

I've attached this figure of SQL Table joining operations. That highlights what I have in mind.

![image](https://github.com/lab-cosmo/equistore/assets/62180584/5fc0c64e-e327-47f9-8b56-9525d376c0f3)

**Comment by agoscinski** [comment]
SQL's join is for merging columns with set rules on the rows (the rule specified by 'inner', 'outer'). We want to merge rows with rules on the columns (here I mean the rows of Labels). Because of that difference we stumble on the problem of duplicates in the rows. That we solve by creating a new label column
https://github.com/lab-cosmo/equistore/blob/ceb892caf456e00e8f76c18791031fdd49828e97/python/equistore-operations/equistore/operations/join.py#L35-L44

As we discussed on slack the functionality in SQL is a bit different, it allows to create `null` values in the table and by that they can joins much simpler. Supporting NaNs as label entries introduces quite a lot of complexity and seems also not really needed for the moment. As soon as one column name mismatch exists (one column is in tensor map A but not in B), the remaining rows cannot be joined as the they would need to be using the nulls as in SQL. For that we use the solution.
https://github.com/lab-cosmo/equistore/blob/ceb892caf456e00e8f76c18791031fdd49828e97/python/equistore-operations/equistore/operations/join.py#L46-L56

I think the equistore join implementation is now an outer join implementation on the Labels of the block, but ignores the tensor map keys completely. This would need to be extended in the current implementation. Secondly one could then add an inner join version of this.

 Having said all this, I am not sure how intuitive it is to use the same nomenclature as SQL. I always found the the SQL join wording confusing and now if you know SQL join, you will get confused because we do stuff different enough. Using just set terms "union", "intersect" seems much more intuitive to me.

--- 
Maybe it is worth it to look at pandas `merge` function (a more versatile join) to get some inspiration. The argument `how ` (inner, outer, left, right) and `on` (for the key to match) argument are used specify how to merge the tables https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html
With our `axis` argument in `operations.join` we have already an `on`. So we could add the argument `how` that specifies how we want to join. 

If you look at the argument `indicator` in the pandas merge API, this is basically our "tensor" solution which we add to differ between the TensorMaps. To deal with duplicates in the columns they also offer to pass `suffixes` a list of strings that are used to append the new column names that have the same name.

**Comment by bananenpampe** [comment]
Hello, this would be the intended API:

```python

def join_block() -> equistore.TensorBlock:
    
    """
    joins tensorblock along a given dimension

    handles logic of checking that samples or properties match ... etc
    """



def outer_join(maps: List[equistore.TensorMap], axis: str) -> equistore.TensorMap:
    
    """
    Outer join of TensorMaps.

    joins all blocks from TensorMaps with identical keys, along given dimension

    new blocks are returned in a new TensorMap.
    blocks with unique keys are kept as they are and are part of the new TensorMap

    
    ----------------------------------------------------

    Simple case, two tensor maps:
    outer_join(TensorMap_A, TensorMap_B) -> TensorMap_out


    ----------------------

    TensorMap_A:

    TensorMap with 2 blocks
    keys: species_center
                1
                8
    
    
    TensorMap_A.block(species_center=1):
    TensorBlock
        samples (100)

    TensorMap_A.block(species_center=8):
    TensorBlock
        samples (100)
    
     
    TensorMap_B:
    TensorMap with 2 blocks
    keys: species_center
                6
                8

    TensorMap_A.block(species_center=6):
    TensorBlock
        samples (100)

    TensorMap_A.block(species_center=8):
    TensorBlock
        samples (100)

    returns:
    
    TensorMap_out:
    TensorMap with 3 blocks
    keys: species_center
                1
                6
                8

    TensorMap_out.block(species_center=1):
    TensorBlock
        samples (100)
    
    TensorMap_out.block(species_center=6):
    TensorBlock
        samples (100)

    TensorMap_out.block(species_center=8):
    TensorBlock
        samples (200)

    ----------------------------------------------------

    

    int_tensorblock_holder = {}

    for map in tensormaps:
        for key, block in map.items():
            if key in int_tensorblock_holder:
                int_tensorblock_holder[key].append(block)
            else:
                int_tensorblock_holder[key] = [block]
    
    new_keys = []
    new_blocks = []

    for key, blocks in int_tensorblock_holder.items():
        new_blocks.append(join_block(blocks))
    
    
    TensorMap_out = TensorMap(new_keys, new_blocks)


    """


def inner_join(maps: List[equistore.TensorMap], axis: str) -> equistore.TensorMap:
    
    """
    Inner join of TensorMaps.

    Returns the joined Tensorblocks of the intersecting keys of all the TensorMaps

    new blocks are returned in a new TensorMap.

    
    ----------------------------------------------------

    Simple case, two tensor maps:
    inner_join(TensorMap_A, TensorMap_B) -> TensorMap_out


    ----------------------

    TensorMap_A:

    TensorMap with 2 blocks
    keys: species_center
                1
                8
    
    
    TensorMap_A.block(species_center=1):
    TensorBlock
        samples (100)

    TensorMap_A.block(species_center=8):
    TensorBlock
        samples (100)
    
     
    TensorMap_B:
    TensorMap with 2 blocks
    keys: species_center
                6
                8

    TensorMap_A.block(species_center=6):
    TensorBlock
        samples (100)

    TensorMap_A.block(species_center=8):
    TensorBlock
        samples (100)

    returns:
    
    TensorMap_out:
    TensorMap with 1 block
    keys: species_center
                8

    TensorMap_out.block(species_center=8):
    TensorBlock
        samples (200)

    ----------------------------------------------------

    

    int_tensorblock_holder = {}

    for map in tensormaps:
        for key, block in map.items():
            if key in int_tensorblock_holder:
                int_tensorblock_holder[key].append(block)
            else:
                int_tensorblock_holder[key] = [block]
    
    new_keys = []
    new_blocks = []

    for key, blocks in int_tensorblock_holder.items():
        if len(blocks) == len(tensormaps):
            new_blocks.append(join_block(blocks))
            new_keys.append(key)
    
    
    if len(new_keys) == 0:
        raise ValueError("No common keys found in TensorMaps") ?
    TensorMap_out = TensorMap(new_keys, new_blocks)


    """


```

**Comment by PicoCentauri** [comment]
The join that we currently have would be the new `inner_join`?

**Comment by bananenpampe** [comment]
Inner_join might remove blocks therefore I think we should leave the current behaviour as the default, so that it is not unexpected and it throws an error when  the keys do not match exactly.
And then the error message could point the user to use "inner" or "outer"

**Comment by Luthaf** [comment]
Instead of different functions, should we make this an argument to join? `different_keys="error"`, `different_keys="intersect"`, `different_keys="union"` or something like this?

**Comment by bananenpampe** [comment]
I agree, I would name it  "matching_keys"/"on_matching_keys"=["inner","outer","error"]

**Comment by PicoCentauri** [comment]
But, only specifying this for the `keys` is not enough, right? You can also specify how to handle the other Labels object in the block. Imagine I want to join along `properties`. I could define how want to handle the other Labels: `keys` AND `samples`. I can do something like:

join `properties`: `different_keys = "intersect"` +`different_samples = "intersect"` or
join `properties`: `different_keys = "intersect"` +`different_samples = "error"`

The current implementation is

join `properties`: `different_keys = "error"` +`different_samples = "error"`

**Comment by Luthaf** [comment]
If you think of a TensorMap as a sparse tensor, the default behavior should be `different_keys="union"` and `different_samples="union"`. That's what the current implementation of join does, though `keys_to_properties`.
